## Art- Database Continued - Descriptions

In [267]:
# import libraries

import pandas as pd
import numpy as np
import requests
import warnings
import matplotlib.pyplot as plt
import seaborn as sns
import re
from transformers import pipeline
from IPython.display import Image, display
import math

# set options
pd.set_option('display.max_columns', None)
pd.set_option('display.max_colwidth',None)
pd.set_option('display.max_rows', 100)
import warnings 
  

warnings.filterwarnings('ignore') 

In [269]:
#load to dataframe
df = pd.read_csv('objects.csv')

dropped_cols = ['contextualtextcount', 'accesslevel', 'dateoflastpageview', 'totaluniquepageviews', 'markscount',
                'publicationcount', 'totaluniquepageviews', 'contact', 'colorcount', 'rank', 'id', 'state', 
                'verificationleveldescription', 'images','imagecount', 'worktypes', 'imagecount', 'classificationid',
                'standardreferencenumber', 'signed', 'relatedcount', 'verificationlevel', 'titlescount',
                'peoplecount', 'lastupdate', 'periodid', 'edition', 'lendingpermissionlevel', 'colors', 'groupcount',
                'dateoffirstpageview', 'datebegin', 'exhibitioncount', 'imagepermissionlevel', 'mediacount', 'techniqueid', 
                'seeAlso', 'details', 'videos', 'audio']

df.drop(dropped_cols, axis = 1, inplace = True)

## Grab just the records with descriptions

In [271]:
description_df = df[~df['description'].isna()]
print(len(description_df))
display(description_df[['description']][:100])

36688


description
0                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         Corresponds to negative BRLF.457.10
1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     

## There are 36688 records with descriptions. Now lets remove records with formated descriptions


In [273]:
description_df =description_df[~description_df.description.str.contains('|'.join(['Corresponds to negative',
                        'Corresponds with negative', "Postcard", 'Corresponds to slide', 'in paper portfolio cover'  
                          'Small-format print', 'Two photographs printed on a single sheet',  'Obv.: aegis\r\nRev.:'                         ]),case=False )]
print(len(description_df))
display(description_df[['description']][:100])

35889


description
9                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             illustrates April, corresponding zodiac sign (cancer) at top center
10                                                                                                                                                                                                                                                                                                                                                

## Simple Sentiment Analysis
### try on description field

In [276]:
# Initialize the sentiment analysis pipeline
model_name = "distilbert/distilbert-base-uncased-finetuned-sst-2-english"
sentiment_pipeline = pipeline("sentiment-analysis", model=model_name)

def clean_text(text):
    """Preprocess text by removing leading numbers/special characters"""
    if pd.isna(text) or str(text).strip() == "":
        return ""
    
    text = str(text)
    # Remove leading numbers followed by punctuation (like "24) ")
    text = re.sub(r'^\d+[\)\.,]\s*', '', text)
    # Remove any other leading special characters
    text = re.sub(r'^[^a-zA-Z0-9"\']+', '', text)
    return text.strip()

def get_sentiment(text):
    try:
        cleaned_text = clean_text(text)
        if not cleaned_text:  # Skip empty text after cleaning
            return None
        
        # Truncate very long texts to model's max length (512 tokens)
        max_length = 500  # Leaving some room for tokenization
        if len(cleaned_text) > max_length:
            cleaned_text = cleaned_text[:max_length]
        
        result = sentiment_pipeline(cleaned_text)[0]
        return result['label']
    except Exception as e:
        print(f"Error processing text: {text}. Error: {e}")
        return None

# Apply to DataFrame
#description_df['desc_sentiment'] = description_df['description'].apply(get_sentiment)

# For sentiment with score
def get_sentiment_with_score(text):
    try:
        cleaned_text = clean_text(text)
        if not cleaned_text:
            return None, None
        
        max_length = 500
        if len(cleaned_text) > max_length:
            cleaned_text = cleaned_text[:max_length]
        
        result = sentiment_pipeline(cleaned_text)[0]
        return result['label'], result['score']
    except Exception as e:
        print(f"Error processing text: {text}. Error: {e}")
        return None, None

description_df[['desc_sentiment', 'sentiment_score']] = description_df['description'].apply(
    lambda x: pd.Series(get_sentiment_with_score(x))
)



Device set to use mps:0


In [210]:
# Display results
display(description_df[['description', 'desc_sentiment', 'sentiment_score']][:50])

description  \
9                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         illustrates April, corresponding zodiac sign (cancer) at top center   
10                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              

## Commentary Analysis
## Grab just the records with commentary, there are 2424 of them

In [278]:
commentary_df = df[~df['commentary'].isna()]
print(len(commentary_df))
display(commentary_df[['commentary']][:100])

2424


commentary
148                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      The patka as an item of clothing has a long history but found its most elegant expression in the age of the Mughals. Examples of it can be seen in almost every portrait of Mughal emperors and courtiers from the late 16th century to the 19th century. The patka was a long, narrow, textile associated with courtly dress. The patka was worn about the wearer's waist, over top a long shirt.
154                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      Index 15 October 1953. For index reference, see Finding Aid [nrs.harvard.edu/urn-3:HUAM:art00006].
174                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                Index 11 May 1957. For index reference, see Finding Aid [nrs.harvard.edu/urn-3:HUAM:art00006].  See related oil (mural) Composition Concrete, 1957 (Davis Catalogue Raisonné‚ no. 1701).
175                                                                                                                                                                                                              

## Remove entries that start with 'Index", this leaves 1258 records

In [280]:
commentary_df = commentary_df[~commentary_df.commentary.str.contains('Index')]
print(len(commentary_df))
display(commentary_df[['commentary']][:100])
commentary_df.to_csv('comments.csv')

1258


commentary
148                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            The patka as an item of clothing has a long history but found its most elegant expression in the age of the Mughals. Examples of it can be seen in almost every portrait of Mughal emperors and courtiers from the late 16th century to the 19th century. The patka was a long, narrow, textile associated with courtly dress. The patka was worn about the wearer's waist, over top a long shirt.
270                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        The use of these types of disks, sometimes called "dischi sacri" ("sacred disks"), is not entirely understood. Several are known, found in Italy, particularly in the south and around ancient Taras/Tarentum (Taranto). Some are in relief, but most have the symbols impressed into the disk.\r\n\r\nThe disks often bear many of the same motifs, some arranged in similar configurations from object to object. The objects may represent divinities--some common associations like Zeus with the thunderbolt, Hephaistos with the tongs, Poseidon with the trident, Hermes with the caduceus, Apollo with the lyre, and Athena with the owl are very clear--but many are in fact everyday items that are not necessrily associated with any particular god or goddess. They may, therefore, represent a mixture of sacred and profane objects, a selection of offerings, or as Prof. McDaniel suggested in his 1924 article about these disks, items loosely associated with the city of Tarentum. Favoring the lat

## With only 1258 records we could likely read them all, lets try sentiment analysis to see if we are able to pick anything up

In [216]:
commentary_df[['com_sentiment', 'com_score']] = commentary_df['commentary'].apply(
    lambda x: pd.Series(get_sentiment_with_score(x))
)

In [217]:
# Display results
display(commentary_df[['commentary', 'com_sentiment', 'com_score']][:100])

commentary  \
148                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            The patka as an item of clothing has a long history but found its most elegant expression in the age of the Mughals. Examples of it can be seen in almost every portrait of Mughal emperors and courtiers from the late 16th century to the 19th century. The patka was a long, narrow, textile associated with courtly dress. The patka was worn about the wearer's waist, over top a long shirt.   
270                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        The use of these types of disks, sometimes called "dischi sacri" ("sacred disks"), is not entirely understood. Several are known, found in Italy, particularly in the south and around ancient Taras/Tarentum (Taranto). Some are in relief, but most have the symbols impressed into the disk.\r\n\r\nThe disks often bear many of the same motifs, some arranged in similar configurations from object to object. The objects may represent divinities--some common associations like Zeus with the thunderbolt, Hephaistos with the tongs, Poseidon with the trident, Hermes with the caduceus, Apollo with the lyre, and Athena with the owl are very clear--but many are in fact everyday items that are not necessrily associated with any particular god or goddess. They may, therefore, represent a mixture of sacred and profane objects, a selection of offerings, or as Prof. McDaniel suggested in his 1924 article about these disks, items loosely associated with the city of Tarentum. Favoring t

In [218]:
display(commentary_df.loc[[5993, 24419, 24934, 11545]][['title', 'commentary', 'com_sentiment', 'com_score']])

title  \
5993                 Untitled (Portfolio of Twelve Prints)   
24419                                            Portfolio   
24934                                     Giacomo Leopardi   
11545  Portrait of Hilaire Germain Edgar Degas (1834-1917)   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       commentary  \
5993                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            As in much of his other work, Richard Prince explores issues of masculinity in this portfolio of twelve lithographs. Each print comprises a smattering of texts and images: some fragmentary, some fully realized, some drawn on the stone, some transferred from other sources. Prince's hand-written texts tell jokes, the kinds of raunchy, yet classic jokes allegedly told among men. The prints are an extension of his painting practice. In the late 1980s, Prince began a series of joke paintings that critics have related to the humor and cartoon imagery featured in men's magazines like Playboy. In fact, the portfolio box that contains the prints the prints is decorated with a Playboy bunny on its cover. Yet although Prince deploys such iconic signifiers of masculinity as the bunny and dirty traveling salesmen jokes, the picture of masculinity he conjures

In [219]:
import asyncio
from playwright.async_api import async_playwright
from bs4 import BeautifulSoup

async def scrape_page(url):
    '''
    A function to webscrape primary image off of the Harvard collections 
    website that will work in a Jupyter Notebook.
    To run: await scrape_page(url)
    
    Args
        url(url): The url of the page to scrape

    Returns
        image_url(url): The first image (url) of the page scraped
    '''
    async with async_playwright() as p:
        # Launch a headless browser
        browser = await p.chromium.launch(headless=True)
        page = await browser.new_page()

        # Navigate to the page
        await page.goto(url)  # Replace with your URL

        # Get the page content
        html_content = await page.content()

        # Parse the HTML with BeautifulSoup
        soup = BeautifulSoup(html_content, 'html.parser')

        # Extract image URLs
        image_urls = []
        for img_tag in soup.find_all('img'):
            src = img_tag.get('src')
            if src and src.startswith("https://nrs.hvrd.art/urn-3:HUAM"):
                src = src.replace('&amp;', '&')
                image_urls.append(src)
            
            
        return image_urls[0]

        # Close the browser
        await browser.close()


In [246]:
i = commentary_df.loc[24419, 'url']  # Get URL directly

try:
    new_image_url = await scrape_page(i)  # Await the scrape
    commentary_df.loc[24419, 'primaryimageurl'] = new_image_url  # Correct assignment

except Exception as e:
    print(f"Error: {e}")  # Better error handling

huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)


In [264]:


for idx in [5993, 24419, 24934, 11545]:

    # URL of the image
    image_url = commentary_df.loc[idx]['primaryimageurl']  #.values[0]
    if pd.notna(image_url) and str(image_url).strip() not in ["", "NaN", "None"]:
        display(Image(url=image_url, width=300))
        print(image_url)
    else:
        print("No valid image URL found.")

https://nrs.harvard.edu/urn-3:HUAM:INV200624_dynmc


https://nrs.hvrd.art/urn-3:HUAM:INV157845_dynmc?width=256&height=256


https://nrs.harvard.edu/urn-3:HUAM:INV044740_dynmc


https://nrs.harvard.edu/urn-3:HUAM:INV177632_dynmc


In [258]:
open_tanning = 'https://www.dorotheatanning.org/images/originals/92.7.03Portfolioopen836.jpg'
display(Image(url=open_tanning, width=300))